### DISCUSSION TITLE
Mapping Organization Id ( or organization attribute ) to Access Token

### DISCUSSION THREAD

**Discussion by magoogli** [keycloak/keycloak#32364] [Q&A] [answered]
Hi,

I have recently started using the Organizations feature in Keycloak which has been extremely useful in our multi tenant application. 

My only problem at the moment is that by default it seems only the Organization name is added to the access token. Is there any way to map the Organization Id into the token, or at least an Organization attribute? I have had a look at the Organization Membership Mapper, but it doesn't appear to have any settings to configure what attributes of the Organization are added ( or am I missing something? )

In my scenario I have a separate Tenant Micro Service which stores additional information about the Tenants / Organizations and ideally I would like to be able to share some kind of ID between it and Keycloak in order to match Organizations up. At the moment I have to do this based on the name of the organization ( as that is what is contained in the Access Token ) and that is not ideal as the name is changeable in keycloak and It means that I am forced to use GUIDS for the name. 

All I really need is the Id of the Organization to be added to the access token.

**Comment by mzvankovich** [comment]
Hi,
As far as I understand, the organization Id is currenttly added to the access token as a separate claim:
```
{
  ...
  "scope": "phone profile organization email address",
  "organizationId": "a56bea03-5904-470a-b21c-92b7f1069d44",
  "address": {},
  "email_verified": false,
  "roles": [
  ],
  "organization": {
    "orga": {}
  },
  ...
}
```

**Reply by magoogli** [reply]
Hi @mzvankovich If this is the case that would be great. What version did you test this on? I have 25.0.4 installed, but am definitely not getting a separate organizationId claim in my access token - only the organization section.

**Reply by mzvankovich** [reply]
Hello @magoogli, it is possiple to specify that realm throuth the protocol mapper in  the client scope (it should be created for a specific dedicated client's scope):
![Screenshot 2024-09-09 142518](https://github.com/user-attachments/assets/e3a93bf2-b9cc-48b1-9555-1e2fb412e2f1)
![Screenshot 2024-09-09 142553](https://github.com/user-attachments/assets/3de75fc0-ee3a-4bb2-a639-4ea2db2ba097)

Although it work for 25.0.4, seems that it's not working for upcoming versions. I tested it in main branch and it doesn't work, the organisation claim still does't contain orgId. Maybe it's going to be changed.

**Reply by mzvankovich** [reply]
By the way the new organisation alias field is introduced in the upcoming keycloak version and taht field is immutable. That filed is added to organisation claim of the access token.
![Screenshot 2024-09-09 212524](https://github.com/user-attachments/assets/141be1f4-fdf1-4c14-96c2-ee286eeb913c)

**Reply by magoogli** [reply]
Hi @mzvankovich, Thank you so much for all your responses and help! The alias field looks like a good option - if it is immutable that will at least mean we can store it safely in our database and maintain the relationship without the danger of it being changed in keycloak. 

I still feel that the lack of the actual identifier is a bit of an oversight. For example the user identifier as stored in keycloak is present in the sub claim - this is very useful as it can be stored in external systems and always be relied on to match up with the user entity in keycloak. Calling any API in keycloak relating to a user will use this identifier. 

The same kind of logic I think should apply to the organisation. At the moment the keycloak APIs relating to the organisation all use the organisation id in the url - as they should, but it means that you need the id to invoke the urls. At the moment that id is not actually present in the access token so you need to jump through some hoops trying to find out what the id is. For example by first listing all organisations in keycloak then matching up names to get the id so that an api can be invoked. This seems a bit convoluted. If the Id was present in the access token it would make this much simpler.

**Reply by mzvankovich** [reply]
Hello @magoogli,  I've created an appropriate issue #32746 for consideration.

**Reply by sguilhen** [reply]
@mzvankovich @magoogli at the moment storing the org ids for future API calls might not be a consistent approach because, unlike users,  the org ids currently change if the realm is exported / reimported. I can't recall exactly the reason but there was some technical difficulty to maintain the alias across reimports. Also, if the token is long-lived, and the realm is re-imported, the id in the token won't map to a valid org.

Having said that, I can see the point of the API usage being a bit convoluted. The `alias` that was introduced is unique and immutable, and it should be possible to obtain the organization associated with the alias by sending a GET request to `admin/realms/{realm}/organizations?q=alias:myalias` and then fetch the id from there for other calls.

So, I personally see two options here: re-evaluate the import/export logic and see if we can find a way to preserve the ids, or allow the orgs to be referenced by the alias in the URL (IIRC the alias has validations in place to keep it URL friendly). So instead of calling the API using the org UUID, one should be able to call `admin/realms/myrealm/organizations/myorg` for example.

**Reply by magoogli** [reply]
Hi @sguilhen thanks for the information - I was not aware that the ids could actually change so I will definitely bear this in mind. As some background - in our system we have a multi tenant architecture where each tenant has their own database. At the moment I am storing meta information about the tenants ( e.g connection string ) in a database that we manage while using keycloak's organization feature to manage all authentication for the tenant . When a call comes in to our backend it will be authenticated with keycloak and I use the organization information in the token to 'lookup' the tenant in our database so I can get the connection string to use ( there is some caching involved but you get the idea ).  So this 'mapping' of keycloak's organization to our external tenant is critical to the whole system. I'm not sure if this is the best approach - I was toying with the idea of storing all tenant related data as attributes in keycloak, but this does somewhat limit the flexibility of the system and make retrieving and storing tenant data a bit cumbersome.

**Reply by mzvankovich** [reply]
Hi @sguilhen, thank you for the clarification. I also was not aware of the organization id mutability. From my point of view it is better to rely on organisation id as immutable uuid resource identificator, while the alias causes some confusion.

**Reply by magoogli** [reply]
@mzvankovich I feel the same as you. My vote would definitely go to keeping the uuid immutable and accessible in some way in the access token. Using the alias to identify the organization feels a bit like a work around.

**Reply by sguilhen** [reply]
@mzvankovich @magoogli Thanks for the feedback and info about your use cases. I will ask the team to look into the import/export and see if we can make it keep the ids across imports, so that it can be safely included in the token as it is in the PR that was opened. I will hold that PR for the moment, and try to get the import/export sorted first (maybe by the end of the week, let's see)

**Reply by mzvankovich** [reply]
@sguilhen I have one more concern about the muttable organization id. There are still no events for organization #31421, but it will be defenetly introduced, at least I have one related PR #32358. I'm not sure how events are treated now during import/export, but all events are stored id db and they are based on a resource uuid. Moreover, currently I'm buildyng the distributed system and propageting such an organization event (Organiazation Created) further through a message broker. The organization id plays a significant role as correlation id or message id.

**Reply by sguilhen** [reply]
@mzvankovich we've had a discussion and the decision was to fix the org import/export and preserve the ids. It will be inline with other resources in Keycloak and is safer from the perspective that an alias can be assigned to a different org (by removing an existing org and creating a new one with the same alias, for example). I'm also reviewing your PR for the events, my plan is to include it along with the org id changes for Keycloak 26.

**Reply by sguilhen** [reply, marked as answer]
Update: with https://github.com/keycloak/keycloak/pull/33495 we can now include the org ids in the claim. I've used @mzvankovich 's original commit, and added a switch to the organization scope mapper to control whether the id is to be included or not (similar to what we do with attributes). Tests were also added along with doc updates. This will be in Keycloak 26.

Thanks for discussion and the contributions, @mzvankovich and @magoogli !

**Reply by magoogli** [reply]
@sguilhen Thank you so much - I'm excited for the Keycloak 26 release!

**Reply by cervenf** [reply]
H,
how the mapper should be setup, please?
I created mapper like this:

![image](https://github.com/user-attachments/assets/061882f5-86f9-427f-9a0c-82245a5d3645)


but organization is not included into token.
Do I need setup something else?
I'm using Keycloak 26.0.6.
Thank you.

**Reply by mzvankovich** [reply]
Hello, specify Claim JSON Type.
![image](https://github.com/user-attachments/assets/5cd54837-37dd-49af-9b9f-f04040acdf1b)

**Reply by ognjen-it** [reply]
If you disable default **organization** client scope (set to optional instead of deafult) from client that you are using and add organization mapper in your custom client scope, the claim will not be visible. Tested 26.0.8

**Comment by mirwaisx** [comment]
I had similar problem, and even after doing all this: https://www.keycloak.org/docs/latest/server_admin/#_mapping_organization_claims_ the token did not show up.
It seems that it will only trigger the right setup of organisation when you create a **fresh Realm with Organisation enabled**. I had an old realm and after enabling the Organisation in the realm, I could still create organisations and organisation mapper but the organisation claim does not appear, also browser flow+ first Broker flow do not get updated.
So it is advisable to create a new realm and enable organisation, then all those settings will be done correctly by keycloak.
Version used: 26.1.2

**Reply by sguilhen** [reply]
Yes, that assessment is correct @mirwaisx . It was a deliberate decision to change the org flows for new realms only because we felt that automatically changing existing flows, which might have been heavily customized by Keycloak admins, could be undesirable. So we limited the functionality to new realms only.

One can still manually setup the conditional organization flows in their existing realms - basically add the same constructs in the same place/order we do for new realms. Once the flows are updated the org authenticator should start working.

**Reply by Tim-Schwalbe** [reply]
@sguilhen can you provide any details how to setup the organisation flow in already existing realms? Maybe export/import them or something easy?

**Reply by Tim-Schwalbe** [reply]
It looks like it's just this and then bind it to clients authentification.

![image](https://github.com/user-attachments/assets/a9629c6c-bf07-41d5-b95c-9cc7e8620095)

![image](https://github.com/user-attachments/assets/9e81ee0b-b3a0-43b7-96fa-082ca38463ce)

**Comment by pedroalves91** [comment]
If I assign an organization to a user it shows in the token, however if I assign two organizations to a user, none of them shows up. Is there a way to show all organizations in the token?

**Reply by ognjen-it** [reply]
You should create new scope organization:*, add to client that you use to authorize user and set as default.
It sould work out of the box

**Comment by touqeershafi** [comment]
is there a way that we allow 1 specific realm role to access any organization without the membership of that organization? through the following scope `organization:org-1` ?
